packages/MapFactory/src/DTK_C_API_Map.hpp

```diff
@@ -0,0 +1,331 @@
+
+        // Copy the transferred field back to the original target layout.
+        Kokkos::deep_copy( Kokkos::subview( target_field.dofs, Kokkos::ALL, 0 ),
+                           target_field_copy );
+
+        // Push the data to the target.
+        _target.pushField( target_field_name, target_field );
+    }
+
+    UserApplication<double, SourceMemSpace> _source;
+    UserApplication<double, TargetMemSpace> _target;
+    std::unique_ptr<NearestNeighborOperator<map_device_type>> _map;
+};
+
+//---------------------------------------------------------------------------//
+// Execution space validation.
+bool validateExecutionSpace( DTK_ExecutionSpace space )
+{
+    switch ( space )
+    {
+    case DTK_SERIAL:
+#if defined( KOKKOS_HAVE_SERIAL )
+        return true;
+        break;
+#else
+        return false;
+        break;
+#endif
+
+    case DTK_OPENMP:
+#if defined( KOKKOS_HAVE_OPENMP )
+        return true;
+        break;
+#else
+        return false;
+        break;
+#endif
+
+    case DTK_CUDA:
+#if defined( KOKKOS_HAVE_CUDA )
+        return true;
+        break;
+#else
+        return false;
+        break;
+#endif
+    default:
+        return false;
+        break;
+    }
+}
+
+// Memory space validation.
+bool validateMemorySpace( DTK_MemorySpace space )
+{
+    switch ( space )
+    {
+    case DTK_HOST_SPACE:
+#if defined( KOKKOS_HAVE_SERIAL ) || defined( KOKKOS_HAVE_OPENMP )
+        return true;
+        break;
+#else
+        return false;
+        break;
+#endif
+    case DTK_CUDAUVM_SPACE:
+#if defined( KOKKOS_HAVE_CUDA )
+        return true;
+        break;
+#else
+        return false;
+        break;
+#endif
+    default:
+        return false;
+        break;
+    }
+}
+
+// Map space validation.
+bool validateMapSpaces( DTK_ExecutionSpace map_space,
+                        DTK_MemorySpace source_space,
+                        DTK_MemorySpace target_space )
+{
+    return validateExecutionSpace( map_space ) &&
+           validateMemorySpace( source_space ) &&
+           validateMemorySpace( target_space );
+}
+
+//---------------------------------------------------------------------------//
+// Create a map.
+DTK_Map *createMap( DTK_ExecutionSpace map_space, MPI_Comm comm,
+                    DTK_UserApplicationHandle source,
+                    DTK_UserApplicationHandle target )
+{
+    DTK_MemorySpace src_space =
+        reinterpret_cast<DataTransferKit::DTK_Registry *>( source )->_space;
+    DTK_MemorySpace tgt_space =
+        reinterpret_cast<DataTransferKit::DTK_Registry *>( target )->_space;
+
+    DTK_INSIST( validateMapSpaces( map_space, src_space, tgt_space ) );
+
+    DTK_Map *map;
+    switch ( map_space )
+    {
+    case DTK_SERIAL:
+        switch ( src_space )
+        {
+        case DTK_HOST_SPACE:
+            switch ( tgt_space )
+            {
+            case DTK_HOST_SPACE:
+                map = new DTK_MapImpl<Serial, HostSpace, HostSpace>(
```

Comment:
You need to protect these with preprocessor directives `#if defined( KOKKOS_ENABLE_SERIAL )`

Reply:
good catch